relay/writer/file_util.go

```diff
@@ -0,0 +1,268 @@
+// limitations under the License.
+
+package writer
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	uuid "github.com/satori/go.uuid"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+
+	"github.com/pingcap/dm/pkg/binlog/event"
+	"github.com/pingcap/dm/pkg/binlog/reader"
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/relay/common"
+)
+
+// checkBinlogHeaderExist checks if the file has a binlog file header.
+func checkBinlogHeaderExist(filename string) (bool, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	buff := make([]byte, fileHeaderLen)
+	n, err := f.Read(buff)
+	if n == 0 && err == io.EOF {
+		return false, nil // empty file
+	} else if n != fileHeaderLen {
+		return false, errors.Errorf("binlog file %s has no enough data, only got % X", filename, buff[:n])
+	} else if err != nil {
+		return false, errors.Annotate(err, "read binlog header")
+	}
+
+	if !bytes.Equal(buff, replication.BinLogFileHeader) {
+		return false, errors.Errorf("binlog file %s header not valid, got % X, expect % X", filename, buff, replication.BinLogFileHeader)
+	}
+	return true, nil
+}
+
+// checkFormatDescriptionEventExist checks if the file has a valid FormatDescriptionEvent.
+func checkFormatDescriptionEventExist(filename string) (bool, error) {
+	// FormatDescriptionEvent always follows the binlog file header
+	exist, err := checkBinlogHeaderExist(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "check binlog file header for %s", filename)
+	} else if !exist {
+		return false, errors.Errorf("no binlog file header at the beginning for %s", filename)
+	}
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	// check whether only the file header
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	fs, err := f.Stat()
+	if err != nil {
+		return false, errors.Errorf("get stat for %s", filename)
+	} else if fs.Size() == int64(fileHeaderLen) {
+		return false, nil // only the file header
+	}
+
+	// seek to the beginning of the FormatDescriptionEvent
+	_, err = f.Seek(int64(fileHeaderLen), io.SeekStart)
+	if err != nil {
+		return false, errors.Annotatef(err, "seek to %d for %s", fileHeaderLen, filename)
+	}
+
+	// parse a FormatDescriptionEvent
+	var found bool
+	onEventFunc := func(e *replication.BinlogEvent) error {
+		if e.Header.EventType != replication.FORMAT_DESCRIPTION_EVENT {
+			return errors.Errorf("got %+v, expect FormatDescriptionEvent", e.Header)
+		}
+		found = true
+		return nil
+	}
+
+	eof, err := replication.NewBinlogParser().ParseSingleEvent(f, onEventFunc)
+	if err != nil {
+		return false, errors.Annotatef(err, "parse %s", filename)
+	} else if eof {
```

Comment:
would `eof == true && found == true` happen?

Reply:
in the current implementation of _go-mysql_, should not happen.
I added a `if` to do the check if _go-mysql_ changed someday, we can observe it.

Reply:
if `ParseSingleEvent parsed too many events` is questionable, should we check log pos? it must be right

Reply:
added a check for log pos, and return `true` if `eof == true && found == true`